src/lib/helpers/project-config.ts

```diff
@@ -0,0 +1,170 @@
+  projectConfig: ProjectConfig | undefined,
+): Promise<DirContext | undefined> => {
+  if (knockDirFlag) {
+    return knockDirFlag;
+  }
+
+  if (!projectConfig) {
+    return undefined;
+  }
+
+  const abspath = path.isAbsolute(projectConfig.knockDir)
+    ? projectConfig.knockDir
+    : path.resolve(process.cwd(), projectConfig.knockDir);
+
+  const exists = await fs.pathExists(abspath);
+
+  // TODO: should we be erroring here if the directory does not exist? or is not a directory?
+  // if (!exists) {
+  //   throw new Error(`${projectConfig.knockDir} does not exist`);
+  // }
+  // if (await isDirectory(abspath)) {
+  //   throw new Error(`${projectConfig.knockDir} is not a directory`);
+  // }
+
+  return { abspath, exists };
+};
+
+/**
+ * Finds and reads the knock.json configuration file.
+ * Returns undefined if the file is not found.
+ *
+ * @returns the project configuration, or undefined if not found
+ */
+export const findAndReadProjectConfig = async (): Promise<
+  ProjectConfig | undefined
+> => {
+  const configPath = await findProjectConfig();
+
+  if (!configPath) {
+    return undefined;
+  }
+
+  return readProjectConfig(configPath);
+};
+
+/**
+ * Resource subdirectory names as they appear in the knock directory.
+ */
+
+export const ResourceDirectoriesByType: Record<
+  Exclude<ResourceType, "reusable_step">,
+  string
+> = {
+  workflow: "workflows",
+  guide: "guides",
+  partial: "partials",
+  email_layout: "layouts",
+  message_type: "message-types",
+  translation: "translations",
+} as const;
+
+/**
+ * Resolves the full path to a resource directory within the knock directory.
+ *
+ * @param projectConfig - the project configuration, if available
+ * @param resourceType - the type of resource
+ * @param basePath - optional base path to resolve relative knockDir against (defaults to cwd)
+ * @returns the absolute path to the resource directory
+ */
+export const resolveResourceDir = async (
+  projectConfig: ProjectConfig | undefined,
+  resourceType: ResourceType,
+  basePath: string = process.cwd(),
+): Promise<DirContext> => {
+  if (!projectConfig) {
+    return {
+      abspath: basePath,
+      exists: true,
+    };
+  }
+
+  const absoluteKnockDir = path.isAbsolute(projectConfig.knockDir)
+    ? projectConfig.knockDir
+    : path.resolve(basePath, projectConfig.knockDir);
+
+  const resourceDir =
+    resourceType in ResourceDirectoriesByType
+      ? ResourceDirectoriesByType[
+          resourceType as Exclude<ResourceType, "reusable_step">
+        ]
+      : undefined;
+
+  if (!resourceDir) {
+    throw new Error(`Unknown resource type: ${resourceType}`);
+  }
+
+  const absResourceDirPath = path.resolve(absoluteKnockDir, resourceDir);
+  const exists = await fs.pathExists(absResourceDirPath);
+
+  // TODO: should we be erroring here if the directory does not exist?
+  // if (!(await isDirectory(absResourceDirPath))) {
+  //   throw new Error(`${absResourceDirPath} is not a directory`);
+  // }
```

Comment:
I'm thinking doing something similar here as I commented above - if the provided path exists but not a directory then error. Otherwise return the directory context and let the caller handle it.